mewarpx/mewarpx/sim_control.py

```diff
@@ -0,0 +1,71 @@
+""" Control code used to terminate simulation based on
+a set of user defined criteria
+"""
+
+import logging
+import time
+
+import numpy as np
+import psutil
+
+from pywarpx import _libwarpx
+from mewarpx.mwxrun import mwxrun
+
+# Get module-level logger
+logger = logging.getLogger(__name__)
+
+class SimControl:
+    """
+    Evaluate whether to continue or terminate the simulation based on a
+    set of user defined functions and criteria
+
+    """
+    def __init__(self, criteria, criteria_args, max_loops=None, **kwargs):
+        """
+        Generate and install functions to perform after a step #.
+        Arguments:
+
+        criteria: list of user defined functions that return True or False value
+
+        criteria_args: list of kwargs for each criteria
+        """
+
+        self.crit_list = criteria
+        self.crit_args_list = criteria_args
+        self.max_loops = max_loops
+        self.check_increment = 0
+
+    def add_checker(self, criteria=None, crit_args=None):
+        self.crit_list.append(criteria)
+        if crit_args:
+            self.crit_args_list.append(crit_args)
+        else:
+            self.crit_args_list.append({})
+
+    def check_criteria(self):
+        """
+        Evaluates each criteria in the crit_list
+        returns True is all criteria are True otherwise returns False
+
+        Example Sim run:
+
+        steps_per_loop = 10
+        while (sim_control.check_criteria()):
+            sim.step(steps_per_loop)
+
+        """
+        if self.check_increment > 0:
+            keep_going = False
+            for i, func in enumerate(self.crit_list):
+                keep_going = self.crit_list[i](**self.crit_args_list[i])
+                if not keep_going:
```

Comment:
I would not short-circuit these function evaluations if one returns false. Sometimes we'll want to install functions that both handle diagnostics and tell something to terminate. It'll probably be most straightforward if those all still run on the last time through things.

Reply:
You can keep the same print info by starting with an empty string, then add to it for each criteria that returns False.